### PR TITLE
feat(Combobox): Loading state with skeletons

### LINKED ISSUE #345: ComboBox: loading state
image
https://react-select.com/home#async

_Originally posted by @mayank99 in https://github.com/iTwin/iTwinUI-react/pull/338#discussion_r718504790_

### LINKED ISSUE #680: ComboBox: Resets input value when new data is loaded
It is impossible to make lazy-loaded search using ComboBox.
For example, if you type something and the query is being made with your input value, after new data is loaded input will be cleared because of the `useEffect`.

### LINKED ISSUE #925: ComboBox: loading state
image
https://react-select.com/home#async

_Originally posted by @mayank99 in https://github.com/iTwin/iTwinUI-react/pull/338#discussion_r718504790_

### PR DESCRIPTION
**Pull request by bentleyvk**
Closes iTwin/iTwinUI#925

combobox-skeleton

I am not sure if it is a convenient way for the users to write this code themselves.

## Checklist

- [x] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [ ] Approve test images for new stories
- [x] Add screenshots of the key elements of the component

### CODE REVIEW

**Comment by bentleyvk** [comment]
@mayank99 @veekeys Can you check how it looks?

**Comment by veekeys** [comment]
Was looking at it for some time, but dont have anything useful to say.
This approach does give full control and possibility to customize the bhavior, which is good. 
But at the same time, it is not that intuitive. 
Do you think it would be possible to add `skeleton` or `isLoading` flag to `MenuItem` too?

**Comment by bentleyvk** [comment]
> Do you think it would be possible to add `skeleton` or `isLoading` flag to `MenuItem` too?

Do you mean to add a prop? I don't want to do that, how we will control whether to show description, icon?

**Comment by bentleyvk** [comment]
> > > Do you think it would be possible to add `skeleton` or `isLoading` flag to `MenuItem` too?
> > 
> > 
> > Do you mean to add a prop? I don't want to do that, how we will control whether to show description, icon?
> 
> if `MenuItem` has sublabel set, we show it?

If there already is content why user needs skeleton at all? 😀 Or you want users to pass some dumb data like `<MenuItem skeleton label="Please show something" description="It would be nice to have skeleton for this" icon={<div className="i-want-skeleton" />} />` Or you want to also have props `showDescription`, `showIcon` but that doesn't make sense to have in regular `MenuItem`.

**Comment by veekeys** [comment]
> > > > Do you think it would be possible to add `skeleton` or `isLoading` flag to `MenuItem` too?
> > > 
> > > 
> > > Do you mean to add a prop? I don't want to do that, how we will control whether to show description, icon?
> > 
> > 
> > if `MenuItem` has sublabel set, we show it?
> 
> If there already is content why user needs skeleton at all? 😀 Or you want users to pass some dumb data like `<MenuItem skeleton label="Please show something" description="It would be nice to have skeleton for this" icon={<div className="i-want-skeleton" />} />` Or you want to also have props `showDescription`, `showIcon` but that doesn't make sense to have in regular `MenuItem`.

yeah.. might be weird :/
I guess go this way then.

**Comment by mayank99** [comment]
`MenuItemSkeleton` looks fine to me, but I don't think this is combobox-specific so maybe we should show the example somewhere else, especially because ComboBox is being refactored and there will be a more convenient way to use it. And the PR title should start with `feat(MenuItem)` or `feat(MenuItemSkeleton)`

**Comment by bentleyvk** [comment]
> `MenuItemSkeleton` looks fine to me, but I don't think this is combobox-specific so maybe we should show the example somewhere else, especially because ComboBox is being refactored and there will be a more convenient way to use it. And the PR title should start with `feat(MenuItem)` or `feat(MenuItemSkeleton)`

It was asked for Combobox so I think we should show example how to achieve it in Combobox. If you not sure, I can just wait until your Combobox PR is merged or we could rework a story after that.

**Comment by mayank99** [comment]
> It was asked for Combobox so I think we should show example how to achieve it in Combobox.

Maybe we can provide them with a codesandbox example in the meantime.

Other than the code, we also wanted to be able to show a loading spinner inside the input, so it makes sense to close the issue when both are available.

File: packages/iTwinUI-react/stories/core/ComboBox.stories.tsx
Comment (bentleyvk) [inline packages/iTwinUI-react/stories/core/ComboBox.stories.tsx:502]: Personally I like this one more as it requires less code from the user.

Comment (elephantcatdog) [inline packages/iTwinUI-react/stories/core/ComboBox.stories.tsx:502]: Why did you put the examples in ComboBox.stories.tsx rather than making a new story for skeletons? Is ComboBox the only setting where these would be used?

**Review by elephantcatdog — COMMENTED** [review]
Some things I noticed that are related more to ComboBox and less to Skeletons:
- "Andorra" is written as "AndorrA" (in Combox.tsx)
- After loading, the combobox dropdown gets wider than the initial box. This is visible in the gif you posted. This also happens without the loading.
- Typing in the box before loading is finished will show "No options found", and when loading is finished the search bar is cleared. This might be the intended behavior? I think it could be better if either (a) the combobox waits for loading to be finished before searching and the search query is remembered or (b) searching isn't allowed while the dropdown is loading.

File: packages/iTwinUI-react/stories/core/ComboBox.stories.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/stories/core/ComboBox.stories.tsx:568]: I made two versions for reviewers to choose from 😀 Sorry forgot to mention that in description.

File: packages/iTwinUI-react/stories/core/ComboBox.stories.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/stories/core/ComboBox.stories.tsx:502]: It was a users' request for a loading state in ComboBox so it is easier for the user to copy from our stories.

**Comment by bentleyvk** [comment]
> Some things I noticed that are related more to ComboBox and less to Skeletons:
> 
> * "Andorra" is written as "AndorrA" (in Combox.tsx)

I will create a PR after that so this one wouldn't have unrelated test img changes.

> * After loading, the combobox dropdown gets wider than the initial box. This is visible in the gif you posted. This also happens without the loading.

Yeah, can't do anything about that as data is loaded later so we don't know what is the widest item in there.

> * Typing in the box before loading is finished will show "No options found", and when loading is finished the search bar is cleared. This might be the intended behavior? I think it could be better if either (a) the combobox waits for loading to be finished before searching and the search query is remembered or (b) searching isn't allowed while the dropdown is loading.

Created an issue https://github.com/iTwin/iTwinUI-react/issues/680

File: packages/iTwinUI-react/stories/core/ComboBox.stories.tsx
Comment (elephantcatdog) [reply packages/iTwinUI-react/stories/core/ComboBox.stories.tsx:568]: Ah, okay. I like "Loading" better also.

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (mayank99) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:101]: Shouldn't these be `onShow` and `onHide` to be consistent with our other components?

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (mayank99) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:70]: This will need additional jsdoc to describe the overload behavior

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (mayank99) [inline packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx:68]: nit: you don't need `='true'` in react

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (mayank99) [inline packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx:77]: I think we might want to make `VisuallyHidden` a separate component

File: apps/storybook/src/ComboBox.stories.tsx
Comment (mayank99) [inline apps/storybook/src/ComboBox.stories.tsx:499]: Should we simplify this? I don't see the point of random widths

File: apps/storybook/src/ComboBox.stories.tsx
Comment (mayank99) [inline apps/storybook/src/ComboBox.stories.tsx:548]: maybe move this into a helper so that the logic becomes simpler
```suggestion
        if (isLoading) {
          const _options = await fetchOptions();
          setIsLoading(false);
          setOptions(_options);
        }
```

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx
Comment (mayank99) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx:503]: nit: it's not very obvious that this is testing the overloading logic, maybe rename

```suggestion
it('should accept ReactNode in emptyStateMessage', async () => {
```

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:101]: I am not sure about naming, whether it will be clear what these props do, ofc they can read JSDocs and then it is obvious.

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx:77]: Maybe when we have more usage of it or you are talking about private component?

File: apps/storybook/src/ComboBox.stories.tsx
Comment (bentleyvk) [reply apps/storybook/src/ComboBox.stories.tsx:499]: It's just nicer for me 😀 but if concern is the amount of code, I can make it all the same width.

Comment (mayank99) [reply apps/storybook/src/ComboBox.stories.tsx:521]: yup readability is the concern. maybe we could reduce it a one liner and inline it like so:

```ts
contentWidth={Math.max(25, Math.random() * 60)} // randomize skeleton width
```

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx]: Renamed

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: Removed

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: Added a private component

File: apps/storybook/src/ComboBox.stories.tsx
Comment (bentleyvk) [reply apps/storybook/src/ComboBox.stories.tsx]: Simplified. I don't know about visual testing as it might cause random failures because content changes after some time.

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:193]: I don't know whether it is good to do this but after selecting an item, options array would be just that item and array is changed only on open and you can see that change: one selected item shortly changes to full options list.

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:224]: Apparently `value` type is `string | string[] | number`

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx
Comment (bentleyvk) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx:625]: Previously it was querying from closed menu and causing me errors.

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (mayank99) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:239]: This comment is out of date now. Maybe change to something like "Update filtered options according to input value"

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (mayank99) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:222]: This comment is out of date now because there is no filtering going on

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (mayank99) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: Missed in one spot 😆

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx
Comment (veekeys) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx:640]: I was trying to use `userEvent` for all the click actions

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (mayank99) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:193]: so wait, why was this change needed?

it does look very noticeable that the list is getting filtered before closing

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (veekeys) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:241]: why these needed to be moved?
Also, previously it was focusing on options change.. wont break anything?

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (veekeys) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:327]: Maybe memoize?

File: apps/storybook/src/ComboBox.stories.tsx
Comment (veekeys) [inline apps/storybook/src/ComboBox.stories.tsx:541]: Maybe reload data on every opening? Or that would mess up too much...

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: Upsy, removed

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:193]: I explained in comment why it is needed 😀 I don't see anything switching on closing, this change was made to fix the opening.

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx]: Updated

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx]: Removed it, now it is pretty straight forward what it does

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: Changed

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: Removed

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx:31]: Mentioned that it is only visible to screen readers and accessibility is mentioned on the whole object

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx]: Changed

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:240]: Needed to move because I am using `inputValue` here so it must go after previous block. It is still resetting focus on input.

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx]: Memoized

File: apps/storybook/src/ComboBox.stories.tsx
Comment (bentleyvk) [reply apps/storybook/src/ComboBox.stories.tsx:541]: Tried, it is showing errors in console that it can't read `__originalIndex` of undefined

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (veekeys) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx:22]: I guess string and number? Need to see how its defined in `CssProperties`

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (veekeys) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:275]: ?

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (veekeys) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:193]: I noticed that too, that after selection only 1 item is visible while closing..

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (veekeys) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:68]: Comment kinda confusing, cause `ReactNode` is also string, right?

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx:22]: image

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:68]: Suggestion?

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:193]: Ok, when I turned off dev tools, it is visible for me too

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (veekeys) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:68]: I would say if JSX.Element is provided, then.....
or maybe just React element?

Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx]: Changed

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx]: Ups, removed

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:193]: Changed other `useEffect` and its dep list, should be fixed an issue when closing. But this `useLayoutEffect` is still needed otherwise when data is loaded it will show skeletons -> No options -> options list.

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: Changed to string because otherwise we would need to handle number by ourselves so I left that to the user.

Comment (veekeys) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: But now it is not only percentage, so description needs to be updated

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (veekeys) [inline packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:225]: will it always have the newest `inputValue` ?

Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:225]: It should

File: packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx
Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/Menu/MenuItemSkeleton.tsx]: Changed description

**Review by veekeys — APPROVED** [review]

File: packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx
Comment (mayank99) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:225]: how can you be sure?

Comment (bentleyvk) [reply packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:225]: So `useEffect` will trigger when `options` changes and then takes latest values of everything including `inputValue`

**Review by elephantcatdog — APPROVED** [review]

**Review by mayank99 — APPROVED** [review]

**Pull request merged**
